src/query/parser/promql/resolve_scalars.go

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package promql
+
+import (
+	"fmt"
+	"math"
+
+	"github.com/m3db/m3/src/query/functions/binary"
+
+	pql "github.com/prometheus/prometheus/promql"
+)
+
+var (
+	errNilScalarArg         = fmt.Errorf("scalar expression is nil")
+	errInvalidNestingFetch  = fmt.Errorf("invalid nesting for fetch")
+	errInvalidNestingVector = fmt.Errorf("invalid nesting for vector conversion")
+)
+
+func resolveScalarArgument(expr pql.Expr) (float64, error) {
+	nesting := 0
+	value, err := resolveScalarArgumentWithNesting(expr, &nesting)
+	// On a regular error, return error
+	if err != nil {
+		return 0, err
+	}
+
+	if nesting != 0 {
+		return 0, fmt.Errorf("promql.resolveScalarArgument: invalid nesting %d", nesting)
+	}
+
+	return value, nil
+}
+
+// resolves an expression which should resolve to a scalar argument
+func resolveScalarArgumentWithNesting(expr pql.Expr, nesting *int) (float64, error) {
```

Comment:
instead of int pointer, let's have multiple return values ?

Reply:
Initially tried it with multiple return values and it wasn't working correctly; can look at it again but don't want to spend too long :)